src/shims/x86/ssse3.rs

```diff
@@ -0,0 +1,206 @@
+            | "phsub.d.128" => {
+                let [left, right] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let (which, saturating) = match unprefixed_name {
+                    "phadd.w.128" | "phadd.d.128" => (mir::BinOp::Add, false),
+                    "phadd.sw.128" => (mir::BinOp::Add, true),
+                    "phsub.w.128" | "phsub.d.128" => (mir::BinOp::Sub, false),
+                    "phsub.sw.128" => (mir::BinOp::Sub, true),
+                    _ => unreachable!(),
+                };
+
+                horizontal_bin_op(this, which, saturating, left, right, dest)?;
+            }
+            // Used to implement the _mm_maddubs_epi16 function.
+            // Multiplies corresponding pairs of packed 8-bit unsigned integer
+            // values contained in the first source operand and packed 8-bit signed
+            // integer values contained in the second source operand, add pairs of
+            // contiguous products with signed saturation, and writes the 16-bit sums to
+            // the corresponding bits in the destination.
+            // https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm_maddubs_epi16
+            "pmadd.ub.sw.128" => {
+                let [left, right] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+
+                let (left, left_len) = this.operand_to_simd(left)?;
+                let (right, right_len) = this.operand_to_simd(right)?;
+                let (dest, dest_len) = this.place_to_simd(dest)?;
+
+                assert_eq!(left_len, right_len);
+                assert_eq!(dest_len, left_len / 2);
+
+                for i in 0..dest_len {
+                    let j1 = i.checked_mul(2).unwrap();
+                    let left1 = this.read_scalar(&this.project_index(&left, j1)?)?.to_u8()?;
+                    let right1 = this.read_scalar(&this.project_index(&right, j1)?)?.to_i8()?;
+
+                    let j2 = j1.checked_add(1).unwrap();
+                    let left2 = this.read_scalar(&this.project_index(&left, j2)?)?.to_u8()?;
+                    let right2 = this.read_scalar(&this.project_index(&right, j2)?)?.to_i8()?;
+
+                    let dest = this.project_index(&dest, i)?;
+
+                    let mul1 = i16::from(left1).checked_mul(right1.into()).unwrap();
+                    let mul2 = i16::from(left2).checked_mul(right2.into()).unwrap();
+                    let res = mul1.saturating_add(mul2);
+
+                    this.write_scalar(Scalar::from_i16(res), &dest)?;
+                }
+            }
+            // Used to implement the _mm_mulhrs_epi16 function.
+            // Multiplies packed 16-bit signed integer values, truncate the 32-bit
+            // product to the 18 most significant bits by right-shifting, round the
+            // truncated value by adding 1, and write bits `[16:1]` to the destination.
```

Comment:
Wtf are those operations... this seems so arbitrary.^^

In which sense does adding 1 mean that something has been "rounded"?

What does "bits [16:1]" mean? Can you use Rust subslice syntax instead?

Reply:
They are as arbitrary as intel decided to make them :)

After adding one, the lowest bit is dropped when taking bits `[16:1]` (aka `1..=16`). Dropping the lowest bit is like dividing by 2, so adding 1 before dividing by 2 is like adding 0.5 after the division and truncating the result.

Reply:
Oh, okay. The text talks about rounding before it even mentions taking some of the bits so this wasn't clear at all. So please say something like: Multiplies packed 16-bit signed integer values, and then divides the 32-bit result by 2 (rounding up) by first adding 1 and then taking the bits `1..=16`.

Reply:
It is rounding to nearest rather than rounding up. And we are still taking the highest 18 bits before dividing by 2.

That would leave us:
```rust
// Multiplies packed 16-bit signed integer values, truncate the 32-bit
// product to the 18 most significant bits by right-shifting, and then
// divides the 18-bit value by 2 (rounding to nearest) by first adding
// 1 and then taking the bits `1..=16`.
```

Reply:
"truncate" → "truncates" but looks good otherwise!